.NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs

```diff
@@ -1744,46 +1719,39 @@ private DateTimeResolutionResult ParseWhichWeek(string text, DateObject referenc
             return ret;
         }
 
-        private static DateTimeResolutionResult GetWeekOfMonth(int cardinal, int month, int year, DateObject referenceDate, bool noYear)
+        private DateTimeResolutionResult GetWeekOfMonth(string cardinalStr, int month, int year, DateObject referenceDate, bool noYear)
         {
             var ret = new DateTimeResolutionResult();
-            var value = ComputeDate(cardinal, 1, month, year);
-            if (value.Month != month)
-            {
-                cardinal -= 1;
-                value = value.AddDays(-Constants.WeekDayCount);
-            }
+            var targetMonday = GetMondayOfTargetWeek(cardinalStr, month, year);
+
+            var futureDate = targetMonday;
+            var pastDate = targetMonday;
 
-            var futureDate = value;
-            var pastDate = value;
             if (noYear && futureDate < referenceDate)
             {
-                futureDate = ComputeDate(cardinal, 1, month, year + 1);
-                if (futureDate.Month != month)
-                {
-                    futureDate = futureDate.AddDays(-Constants.WeekDayCount);
-                }
+                futureDate = GetMondayOfTargetWeek(cardinalStr, month, year + 1);
             }
 
             if (noYear && pastDate >= referenceDate)
             {
-                pastDate = ComputeDate(cardinal, 1, month, year - 1);
-                if (pastDate.Month != month)
-                {
-                    pastDate = pastDate.AddDays(-Constants.WeekDayCount);
-                }
+                pastDate = GetMondayOfTargetWeek(cardinalStr, month, year - 1);
             }
 
+            var monthTimex = DateTimeFormatUtil.LuisDate(year, month);
```

Comment:
Let's make sure all timex generation code comes from the generator. Even in a case like this, it could be a wrapper method. Centralizing timex generation will make maintenance easier.

Reply:
Fixed